

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Examples &mdash; DICAST 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/_templates/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/_templates/custom.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Config Files (advanced)" href="../advanced.html" />
    <link rel="prev" title="Config files" href="config.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html">
          

          
            
            <img src="../../_static/dicast_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Get started</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../setup.html">Setup</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../setup/git.html">Clone Git Repository</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../setup/docker.html">Setup docker</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../setup/snakemake.html">Install Snakemake</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../setup/input.html">Folder Structure</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../run.html">Run your analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="one_tool.html">Run one specific tool</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="runscript.html">Use multiple tools at once</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="config.html">Config files</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Examples</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../advanced.html">Config Files (advanced)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../advanced/general_parameters.html">General parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../advanced/mapping_parameters.html">Mapping parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="../advanced/as_parameters.html">AS detection parameters</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Tools</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../tools/tools.html">General Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tools/mapping.html">Mapping tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/bbmap.html">BBMap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/contextmap.html">ContextMap 2.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/crac.html">CRAC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/dart.html">Dart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/gsnap.html">GSNAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/hisat.html">HISAT2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/mapsplice.html">MapSplice 2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/minimap.html">Minimap2</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/segemehl.html">segemehl</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/star.html">STAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/mapping/subjunc.html">Subjunc</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../tools/splicing.html">Splicing tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/asgal.html">ASGAL</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/aspli.html">Aspli</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/eventpointer.html">EventPointer</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/irfinder.html">IRFinder</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/kissplice.html">kisSplice</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/majiq.html">MAJIQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/sgseq.html">SGSeq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/spladder.html">SplAdder</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../tools/splicing/whippet.html">Whippet</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Further information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../further-information/faq.html">FAQ</a><ul class="simple">
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">DICAST</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../run.html">Run your analysis</a> &raquo;</li>
        
      <li>Examples</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/get-started/run/example-files.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>Here we provide the detailed description of possible workflows. We recommend to run analysis using a terminal multiplexer, e.g. tmux or screen.</p>
<div class="section" id="running-multiple-mapping-tools-e-g-star-hisat2-and-bbmap">
<h2>Running multiple mapping tools (E.g., STAR, HISAT2 and bbmap)<a class="headerlink" href="#running-multiple-mapping-tools-e-g-star-hisat2-and-bbmap" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>Make sure you followed the steps described in the <a class="reference internal" href="../setup.html"><span class="doc">setup</span></a> section carefully.</p></li>
<li><p>Before getting started make sure to activate the snakemake conda environment:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><style type="text/css">
span.prompt1:before {
  content: "$ ";
}
</style><span class="prompt1">conda activate snakemake</span>
</pre></div></div><ol class="arabic simple" start="3">
<li><p>Create the <cite>input</cite> folder:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/</span>
<span class="prompt1">mkdir input</span>
</pre></div></div><ol class="arabic simple" start="4">
<li><p>Create the folder structure as in the <cite>sample_output</cite>:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> input</span>
<span class="prompt1">mkdir controldir</span>
<span class="prompt1"><span class="nb">cd</span> controldir</span>
<span class="prompt1">mkdir fastqdir</span>
</pre></div></div><ol class="arabic simple" start="5">
<li><p>Download or copy the genome fasta file into the <cite>input</cite> folder. Dont’t forget to uncompress it. E.g.:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/input</span>
<span class="prompt1">wget http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span>
<span class="prompt1">gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span>
</pre></div></div><ol class="arabic simple" start="6">
<li><p>Download or copy the genome gtf annotation into the <cite>input</cite> folder. Dont’t forget to uncompress it. E.g.:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">wget http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.gtf.gz</span>
<span class="prompt1">gunzip Homo_sapiens.GRCh38.104.gtf.gz</span>
</pre></div></div><ol class="arabic simple" start="7">
<li><p>Download or copy the fastq files you want to align into the <cite>/path/to/DICAST/input/controldir/fastqdir</cite>. Note: we support only paired-end RNA-Seq - fastq files have to be in pairs.</p></li>
<li><p>Go to <cite>/path/to/DICAST/scripts</cite> and edit <cite>config.sh</cite> according to your run (see <a class="reference internal" href="config.html"><span class="doc">How to change your config.sh file</span></a>):</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/scripts</span>
<span class="prompt1">nano config.sh</span>
</pre></div></div><p>In the <cite>config.sh</cite> file edit the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><style type="text/css">
span.prompt2:before {
  content: " ";
}
</style><span class="prompt2">read_length=76</span>
<span class="prompt2">fastaname=Homo_sapiens.GRCh38.dna.primary_assembly.fa</span>
<span class="prompt2">gtfname=Homo_sapiens.GRCh38.104.gtf</span>
</pre></div></div><ol class="arabic simple" start="9">
<li><p>List the mapping tools you want to run:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/scripts/snakemake/</span>
<span class="prompt1">nano snakemake_config.yaml</span>
</pre></div></div><p>In the <cite>snakemake_config.yaml</cite> file edit the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt2">Mapping_tools:</span>
<span class="prompt2">    What_tools_to_run: &#39;star, hisat, bbmap&#39;</span>
</pre></div></div><ol class="arabic simple" start="10">
<li><p>In the <cite>/path/to/DICAST/scripts/snakemake/</cite> folder run:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">snakemake -j <span class="m">1</span> -d /path/to/DICAST/input -s Snakefile -c snakemake_config.yaml</span>
</pre></div></div><p>This command will start the mapping tools indicated in the <cite>snakemake_config.yaml</cite> (E.g. STAR, HISAT2 and bbmap).</p>
<p>First, the pipeline will build all necessary dockers. Second, in will create a <cite>/path/to/DICAST/index</cite> folder and put the results of indexing. Finally, the pipeline will create a <cite>/path/to/DICAST/output</cite> folder with the alignment results inside the dedicated folders (e.g., star-output, hisat-output, bbmap-output).</p>
</div>
<div class="section" id="running-multiple-alternative-splicing-event-detection-tools-e-g-majiq-and-whippet">
<h2>Running multiple alternative splicing event detection tools (E.g., MAJIQ and Whippet)<a class="headerlink" href="#running-multiple-alternative-splicing-event-detection-tools-e-g-majiq-and-whippet" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>Make sure you followed the steps described in the <a class="reference internal" href="../setup.html"><span class="doc">setup</span></a> section carefully.</p></li>
<li><p>Before getting started make sure to activate the snakemake conda environment:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda activate snakemake</span>
</pre></div></div><ol class="arabic simple" start="3">
<li><p>Create the <cite>input</cite> folder:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/</span>
<span class="prompt1">mkdir input</span>
</pre></div></div><ol class="arabic simple" start="4">
<li><p>Create the folder structure as in the <cite>sample_output</cite>:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> input</span>
<span class="prompt1">mkdir controldir</span>
<span class="prompt1"><span class="nb">cd</span> controldir</span>
<span class="prompt1">mkdir fastqdir</span>
<span class="prompt1">mkdir bamdir</span>
</pre></div></div><ol class="arabic simple" start="5">
<li><p>Download or copy the genome fasta file into the <cite>input</cite> folder. Dont’t forget to uncompress it. E.g.:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/input</span>
<span class="prompt1">wget http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span>
<span class="prompt1">gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span>
</pre></div></div><ol class="arabic simple" start="6">
<li><p>Download or copy the genome gtf annotation into the <cite>input</cite> folder. Dont’t forget to uncompress it. E.g.:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">wget http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.gtf.gz</span>
<span class="prompt1">gunzip Homo_sapiens.GRCh38.104.gtf.gz</span>
</pre></div></div><ol class="arabic simple" start="7">
<li><p>Download or copy the genome gff3 annotation into the <cite>input</cite> folder (for MAJIQ). Dont’t forget to uncompress it. E.g.:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">wget http://ftp.ensembl.org/pub/release-104/gff3/homo_sapiens/Homo_sapiens.GRCh38.104.gff3.gz</span>
<span class="prompt1">gunzip Homo_sapiens.GRCh38.104.gff3.gz</span>
</pre></div></div><ol class="arabic simple" start="8">
<li><p>Download or copy the fastq files you want to use into the <cite>/path/to/DICAST/input/controldir/fastqdir</cite>. Note: we support only paired-end RNA-Seq - fastq files have to be in pairs.</p></li>
<li><p>Download or copy the bam files you want to use into the <cite>/path/to/DICAST/input/controldir/bamdir</cite>.</p></li>
<li><p>Go to <cite>/path/to/DICAST/scripts</cite> and edit <cite>config.sh</cite> according to your run (see <a class="reference internal" href="config.html"><span class="doc">How to change your config.sh file</span></a>):</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/scripts</span>
<span class="prompt1">nano config.sh</span>
</pre></div></div><p>In the <cite>config.sh</cite> file edit the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nv">read_length</span><span class="o">=</span><span class="m">76</span></span>
<span class="prompt1"><span class="nv">fastaname</span><span class="o">=</span>Homo_sapiens.GRCh38.dna.primary_assembly.fa</span>
<span class="prompt1"><span class="nv">gtfname</span><span class="o">=</span>Homo_sapiens.GRCh38.104.gtf</span>
<span class="prompt1"><span class="nv">gffname</span><span class="o">=</span>Homo_sapiens.GRCh38.104.gff3</span>
</pre></div></div><ol class="arabic simple" start="11">
<li><p>List the mapping tools you want to run:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">cd</span> /path/to/DICAST/scripts/snakemake/</span>
<span class="prompt1">nano snakemake_config.yaml</span>
</pre></div></div><p>In the <cite>snakemake_config.yaml</cite> file edit the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt2">Alternative_splicing_detection_tools:</span>
<span class="prompt2">   What_tools_to_run: &#39;majiq, whippet&#39;</span>
</pre></div></div><ol class="arabic simple" start="12">
<li><p>In the <cite>/path/to/DICAST/scripts/snakemake/</cite> folder run:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">snakemake -j <span class="m">1</span> -d /path/to/DICAST/input -s Snakefile -c snakemake_config.yaml</span>
</pre></div></div><p>This command will start the mapping tools indicated in the <cite>snakemake_config.yaml</cite> (E.g. MAJIQ, Whippet).</p>
<p>First, the pipeline will build all necessary dockers. Second, the pipeline will create a <cite>/path/to/DICAST/output</cite> folder with the event detecton results inside the dedicated folders (e.g., majiq-output, hisat-output, whippet-output).</p>
<div class="toctree-wrapper compound">
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../advanced.html" class="btn btn-neutral float-right" title="Config Files (advanced)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="config.html" class="btn btn-neutral float-left" title="Config files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    
  <head>
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto+Condensed">
	</head>
	<style>


/* Sidebar header (and topbar for mobile) */
.wy-side-nav-search, .wy-nav-top {
	background: #60cdbf;
}

/* Caption headers for sidebar */
.caption-text {
	color: #13736a;
}

/*headings*/
h1, h2, .rst-content .toctree-wrapper p.caption, h3, h4, h5, h6, legend {
	font-family: "Roboto Condensed","ff-tisa-web-pro","Georgia",Arial,sans-serif;
}

.rst-content .sidebar .sidebar-title {
	font-family: "Roboto Condensed","ff-tisa-web-pro","Georgia",Arial,sans-serif;
}

/* Warning  */
div .admonition.warning {
	background: #ffa8ae;
}
	
.rst-content .warning .admonition-title {
	background: #8b272e;
}

/* Note */
div .admonition.note {
	background: #b9fff4;
}
	
.rst-content .note .admonition-title {
	background: #1ca58f;
}

/* List  */
.rst-content dl:not(.docutils) dt {
	color: #13736a;
	background: #e7f2fa;
	border-top: solid 2px #13736a;
}

/* code block */

div .highlight {
	background: #d6f1e9;
}

.nv {
	color: #8b272e;
}

/* inline (guilabel) */

.rst-content .guilabel {
border: 1px solid #7fbbe3;
background: #e7f2fa;
border-radius: 1px;
}

/*inline code*/

.rst-content code.literal {
color: #13736a;	
}

/* links */
a {
  color: #13736a;
}



	</style>



</body>
</html>